pkg/controllers/nodeclaim/capacityreservation/controller.go

```diff
@@ -0,0 +1,153 @@
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named(c.Name()).
+		WatchesRawSource(singleton.Source()).
+		Complete(singleton.AsReconciler(c))
+}
+
+func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	cpNodeClaims, err := c.cp.List(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+	}
+	cpNodeClaimIndex := lo.SliceToMap(cpNodeClaims, func(nc *karpv1.NodeClaim) (string, *karpv1.NodeClaim) {
+		return nc.Status.ProviderID, nc
+	})
+
+	ncs := &karpv1.NodeClaimList{}
+	if err := c.kubeClient.List(ctx, ncs, client.MatchingLabels{
+		karpv1.NodeRegisteredLabelKey: "true",
+	}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeclaims, %w", err)
+	}
+	updatedNodeClaims := sets.New[string]()
+	var errs []error
+	for i := range ncs.Items {
+		cpNC, ok := cpNodeClaimIndex[ncs.Items[i].Status.ProviderID]
+		if !ok {
+			continue
+		}
+		updated, err := c.syncCapacityType(ctx, cpNC.Labels[karpv1.CapacityTypeLabelKey], &ncs.Items[i])
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if updated {
+			updatedNodeClaims.Insert(ncs.Items[i].Name)
+		}
+	}
+	log.FromContext(ctx).WithValues("NodeClaims", lo.Map(updatedNodeClaims.UnsortedList(), func(name string, _ int) klog.ObjectRef {
+		return klog.KRef("", name)
+	})).V(1).Info("updated capacity type for nodeclaims")
+
+	if len(errs) != 0 {
+		if lo.EveryBy(errs, func(err error) bool { return errors.IsConflict(err) }) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, multierr.Combine(errs...)
+	}
+	return reconcile.Result{RequeueAfter: time.Minute}, nil
+}
+
+// syncCapacityType will update the capacity type for the given NodeClaim. This accounts for the fact that capacity
+// reservations will expire, demoting NodeClaims with capacity type "reserved" to "on-demand".
+func (c *Controller) syncCapacityType(ctx context.Context, capacityType string, nc *karpv1.NodeClaim) (bool, error) {
+	// We won't be able to sync deleting NodeClaims, and there's no real need to either as they're already draining.
+	if !nc.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	// For now we only account for the case where a reserved NodeClaim becomes an on-demand NodeClaim. This does not
+	// account for on-demand NodeClaims being promoted to reserved since that is not natively supported by Karpenter.
+	if capacityType != karpv1.CapacityTypeOnDemand {
+		return false, nil
+	}
+	if nc.Labels[karpv1.CapacityTypeLabelKey] == karpv1.CapacityTypeReserved {
+		stored := nc.DeepCopy()
+		nc.Labels[karpv1.CapacityTypeLabelKey] = karpv1.CapacityTypeOnDemand
+		delete(nc.Labels, cloudprovider.ReservationIDLabel)
+		if err := c.kubeClient.Patch(ctx, nc, client.MergeFrom(stored)); err != nil {
+			return false, fmt.Errorf("patching nodeclaim %q, %w", nc.Name, err)
```

Comment:
Do we handle the NodeClaim NotFound error here?

Reply:
Yeah, we do. We should ignore not found errors.